plan/physical_plan_builder.go

```diff
@@ -179,23 +185,39 @@ func (ds *DataSource) findBestTask(prop *requiredProp) (task, error) {
 	if t != nil {
 		return t, nil
 	}
+
+	oldPropCols := prop.cols
+	if prop.enforced {
+		prop.cols = []*expression.Column{}
+	}
 	t, err := ds.tryToGetDualTask()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	if t != nil {
-		ds.storeTask(prop, t)
+		if prop.enforced {
```

Comment:
we can add a `defer` after line 192 and all the following duplicated code can be removed:
```go
defer func() {
    if err != nil || t == nil {
        return
    }

    if prop.enforced {
        prop.cols = oldPropCols
        t = prop.enforceProperty(t, ds.basePlan.ctx)
    }

    ds.storeTask(prop, t)
}()
```

Reply:
Does it work? `defer` is executed after `return`, so the line `t = prop.enforceProperty(t, ds.basePlan.ctx)` will not be effective for return.

Reply:
yes, you can define the function like this:
```go
findBestTask(prop *requiredProp) (t task, err error)
```

in line 184: s/`t := ds.getTask(prop)`/`t = ds.getTask(prop)`/
in line 193: s/`t, err := ds.tryToGetDualTask()`/`t, err = ds.tryToGetDualTask()`/